src/PostgREST/App.hs

```diff
@@ -44,143 +45,197 @@ import qualified Hasql.Backend             as B
+                            "/" <> cs (tableName table) <>
+                              if Prelude.null canonical then "" else "?" <> cs canonical
+                          )
+                        ] (fromMaybe "[]" body)
+                where
+                  frm = fromMaybe 0 $ rangeOffset <$> range
+
+            "POST" ->
+              case request of
+                Left e -> return $ responseLBS status400 [jsonH] $ cs e
+                Right (selectQuery, mutateQuery, isSingle) -> do
+                  row <- H.maybeEx $ B.Stmt (createWriteStatement selectQuery mutateQuery isSingle echoRequested pKeys isCsv) V.empty True
+                  let (_, _, location, body) = extractQueryResult row
+                      contentLocation = cs fullPath <> "?" <> cs (fromMaybe "" location)
+                  return $ responseLBS status201
+                    [contentTypeH, (hLocation, contentLocation)]
+                    $ if echoRequested then fromMaybe "[]" body else ""
+
+            "PATCH" ->
+              case request of
+                Left e -> return $ responseLBS status400 [jsonH] $ cs e
+                Right (selectQuery, mutateQuery, _) -> do
+                  let q = B.Stmt (createWriteStatement selectQuery mutateQuery False echoRequested [] isCsv) V.empty True
+                  row <- H.maybeEx q
+                  let (_, queryTotal, _, body) = extractQueryResult row
+                      r = contentRangeH 0 (queryTotal-1) (Just queryTotal)
+                      s = case () of _ | queryTotal == 0 -> status404
+                                       | echoRequested -> status200
+                                       | otherwise -> status204
+                  return $ responseLBS s [contentTypeH, r]
+                    $ if echoRequested then fromMaybe "[]" body else ""
+
+            "DELETE" ->
+              case request of
+                Left e -> return $ responseLBS status400 [jsonH] $ cs e
+                Right (selectQuery, mutateQuery, _) -> do
+                  let q = B.Stmt (createWriteStatement selectQuery mutateQuery False False [] isCsv) V.empty True
+                  row <- H.maybeEx q
+                  let (_, queryTotal, _, _) = extractQueryResult row
+                  return $ if queryTotal == 0
+                    then responseLBS status404 [] ""
+                    else responseLBS status204 [("Content-Range", "*/" <> cs (show queryTotal))] ""
+
+            "OPTIONS" -> do
+              let cols = filter (filterCol schema tableN) allCols
+                  body = encode (TableOptions cols pKeys)
+              return $ responseLBS status200 [jsonH, allOrigins] $ cs body
+
+            _ -> notFound
+
+          where
+            tableN = tableName table
+            request = parseRequest schema allRels tableN req [] reqBody
+            range = rangeRequested hdrs
+            pKeys = map pkName $ dbFindPrimaryKeys dbStructure table
+
+        RowE (Locus table filters) ->
```

Comment:
This is a lot of code just to support /endpoint/:id route, most of the code is duplicated from the /endpoint/ routes above